src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1252,6 +1270,30 @@ private async ValueTask<(Stream?, TransportContext?, HttpResponseMessage?)> Conn
                             return (null, null, response);
                         }
                         break;
+
+                    case HttpConnectionKind.SocksTunnel:
+                    case HttpConnectionKind.SslSocksTunnel:
+                        Debug.Assert(_originAuthority != null);
+                        Debug.Assert(_proxyUri != null);
+                        stream = await ConnectToTcpHostAsync(_proxyUri.IdnHost, _proxyUri.Port, request, async, cancellationToken).ConfigureAwait(false);
+
+                        if (string.Equals(_proxyUri.Scheme, "socks5", StringComparison.OrdinalIgnoreCase))
+                        {
+                            await SocksHelper.EstablishSocks5TunnelAsync(stream, _originAuthority.IdnHost, _originAuthority.Port, _proxyUri, ProxyCredentials, cancellationToken).ConfigureAwait(false);
+                        }
+                        else if (string.Equals(_proxyUri.Scheme, "socks4a", StringComparison.OrdinalIgnoreCase))
+                        {
+                            await SocksHelper.EstablishSocks4TunnelAsync(stream, true, _originAuthority.IdnHost, _originAuthority.Port, _proxyUri, ProxyCredentials, cancellationToken).ConfigureAwait(false);
+                        }
+                        else if (string.Equals(_proxyUri.Scheme, "socks4", StringComparison.OrdinalIgnoreCase))
+                        {
+                            await SocksHelper.EstablishSocks4TunnelAsync(stream, false, _originAuthority.IdnHost, _originAuthority.Port, _proxyUri, ProxyCredentials, cancellationToken).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            Debug.Fail("Bad socks version.");
+                        }
+                        break;
```

Comment:
I would move this switch based on the scheme into SocksHelper

Reply:
Moving is fine. Unfortunately switch cannot operation case insensitively.